### PR TITLE
Minor improvements to string rewriter

### PR DESCRIPTION
**Pull request by ajreynol**
Handles simple cases found by --sygus-rr-synth.

Work towards #1180.

One function in theory_strings.cpp had to be modified slightly to account for new cases where the string rewriter is able to rewrite equalities to false.  Unfortunately this led to an indentation change and so it was reformatted, but only minor changes/cleaning were done to that function.

### CODE REVIEW

File: src/theory/strings/theory_strings.cpp [hunk 1]
```diff
@@ -2769,170 +2769,203 @@ bool TheoryStrings::processLoop( std::vector< std::vector< Node > > &normal_form
...
+  Trace("strings-loop") << "Detected possible loop for "
+                        << normal_forms[loop_n_index][loop_index] << std::endl;
+  Trace("strings-loop") << " ... (X)= " << normal_forms[other_n_index][index]
+                        << std::endl;
+
+  Trace("strings-loop") << " ... T(Y.Z)= ";
+  std::vector<Node> vec_t;
+  for (int lp = index; lp < loop_index; ++lp)
+  {
+    if (lp != index) Trace("strings-loop") << " ++ ";
```
Comment (4tXJ7f) [inline src/theory/strings/theory_strings.cpp]: You removed these kinds of debug traces from the other loops, is this needed here? If you remove the debug traces, you can also remove the for loop (see other comments).

File: src/theory/strings/theory_strings.cpp [hunk 2]
```diff
@@ -2769,170 +2769,203 @@ bool TheoryStrings::processLoop( std::vector< std::vector< Node > > &normal_form
...
+  }
+
+  Node split_eq;
+  for (unsigned r = 0; r < 2; r++)
+  {
+    Node t = r == 0 ? normal_forms[loop_n_index][loop_index] : t_yz;
+    split_eq = t.eqNode(d_emptyString);
+    Node split_eqr = Rewriter::rewrite(split_eq);
+    // the equality could rewrite to false
+    if (!split_eqr.isConst())
```
Comment (4tXJ7f) [inline src/theory/strings/theory_strings.cpp]: Assert in the else case that split_eqr is not `true`.

File: src/theory/strings/theory_strings_rewriter.cpp [hunk 3]
```diff
@@ -236,6 +237,17 @@ Node TheoryStringsRewriter::rewriteEquality(Node node)
     }
   }
 
+  // ( len( s ) != len( t ) ) => ( s == t ---> false )
+  // This covers cases like str.++( x, x ) == "a" ---> false
+  Node len0 = NodeManager::currentNM()->mkNode(kind::STRING_LENGTH, node[0]);
+  Node len1 = NodeManager::currentNM()->mkNode(kind::STRING_LENGTH, node[1]);
+  Node len_eq = len0.eqNode(len1);
+  len_eq = Rewriter::rewrite(len_eq);
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/strings/theory_strings_rewriter.cpp]: Actually this is to catch cases where an entailment check for inequality doesn't suffice, for example to show 2*len( x ) != 1, we need to invoke the arithmetic rewrite for integers which finds that a divisibility constraint is violated, where notice that neither 2*len( x ) > 0 or 2*len( x )<0 is entailed.

**Comment by ajreynol** [comment]
Anything remaining for this pull request?  A couple of the comments are still showing that were addressed.

File: src/theory/strings/theory_strings_rewriter.cpp [hunk 4]
```diff
@@ -1632,9 +1613,19 @@ Node TheoryStringsRewriter::rewriteContains( Node node ) {
     }else{
       if (s.size() == 0)
       {
-        Node ret =
-            NodeManager::currentNM()->mkNode(kind::EQUAL, node[0], node[1]);
-        return returnRewrite(node, ret, "ctn-emptystr");
+        Node len1 =
+            NodeManager::currentNM()->mkNode(kind::STRING_LENGTH, node[1]);
+        if (checkEntailArith(len1, true))
```
[the remark this answers is not in the document]
Comment (4tXJ7f) [reply src/theory/strings/theory_strings_rewriter.cpp]: My comment here doesn't make sense. The code looks fine.

**Pull request merged**